Update GemTypo to use the -/_ variation detection

rietta opened the pull request:
This update changes the method used by GemTypo from the Levenshtein distance
comparison with popular Gems with 10 million or more downloads to a simpler
check that compares the proposed Gem with all of its - and _ characters removed
to all existing Gems with their - and _ characters removed.

This comparison is performed completely within PostgreSQL and an index is added
to ensure that a table scan is not necessary. This means that this typo protection
can be enforced against any other already published Gem rather than just popular
ones with 10 million or more downloads.

To support the suitability of this new approach, we conducted an analysis of the malicious gems published in the Reversing Labs data set from February 2020. A rake task was used to use the GemTypo class to check each known malicious gem to for matches. Timing data was collected and written out to a CSV file, which was then loaded into spreadsheet to compute the relevant statistics.

## Timing Analysis for Typo Detection Using New Method

The code for collecting the data was:
```ruby
baddies.each do |malicious_gem|
  start_time = Time.now
  typo_check = GemTypo.new(malicious_gem)
  typo_check.protected_typo?
  end_time = Time.now

  puts [malicious_gem, typo_check.protected_gem, end_time - start_time].to_csv
end
```
With a sample size of 725 database checks, the timing in seconds for checks resulting in:

- Min Seconds 0.0087
- Max Seconds  0.2499
- Average Seconds 0.1221



Please see this attached XLSX for details.

[Malicious Gem Detection Timing Statistics Based on New Method and Reversing Labs List of Baddies.xlsx](https://github.com/rietta-inc/rubygems.org/files/4559836/Malicious.Gem.Detection.Timing.Statistics.Based.on.New.Method.and.Reversing.Labs.List.of.Baddies.xlsx)


## Time Required to Update PostgreSQL Schema to Add Index to rubygems Table

```
time be rake db:migrate
== 20200429005140 AddDashUnderscoreTypoDetectionIndexToRubygems: migrating ====
-- execute("DROP INDEX IF EXISTS dashunderscore_typos_idx;")
   -> 0.0012s
-- execute("CREATE INDEX dashunderscore_typos_idx ON rubygems (regexp_replace(upper(name), '[_-]', '', 'g'));")
   -> 0.4861s
== 20200429005140 AddDashUnderscoreTypoDetectionIndexToRubygems: migrated (0.4873s) 


real    0m2.090s
user    0m1.303s
sys     0m0.203s
```

## Explanation of Query Used in New protected_typo? check in GemTypo class
When `protected_typo?` is called, the updated GemTypo class will perform a single query intended to obtain from PostgreSQL the first detected match with minimum data fetched. It does this by calling ` match = matched_protected_gem_names.select(:id, :name).first`

which translates to to following SQL (when called on the proposed typo delayed-job_active_record):

```sql
SELECT "rubygems"."id", "rubygems"."name" FROM "rubygems" WHERE (name != 'delayed-job_active_record' AND regexp_replace(upper(name), '[_-]', '', 'g') = regexp_replace(upper('delayed-job_active_record'), '[_-]', '', 'g')) ORDER BY "rubygems"."id" LIMIT 1
```

The following EXPLAIN analysis shows how PostgreSQL matches this query:
```
Sort  (cost=1350.93..1353.07 rows=854 width=18)
Sort Key: id
->  Bitmap Heap Scan on rubygems  (cost=23.04..1309.35 rows=854 width=18)
Recheck Cond: (regexp_replace(upper((name)::text), '[_-]'::text, ''::text, 'g'::text) = 'DELAYEDJOBACTIVERECORD'::text)
Filter: ((name)::text <> 'delayed-job-active_record'::text)
->  Bitmap Index Scan on dashunderscore_typos_idx  (cost=0.00..22.83 rows=854 width=0)
Index Cond: (regexp_replace(upper((name)::text), '[_-]'::text, ''::text, 'g'::text) = 'DELAYEDJOBACTIVERECORD'::text)
```

## Finally, a Summary of the Data Analysis Basis for Choosing this Method as Appropriate for Detecting Many Malicious Typo Squats 
My April 2020 analysis of the typo squatting attack in February of 2020 found that
none of the target Gems were protected at the time and the vast majority of the
malicious Gem names would have been prevented by this check if it had been in place
at the time. Please see https://speakerdeck.com/rietta/securing-the-open-source-software-supply-chain
slides 22 and 23 for the publication of my findings at this time.

![RubyGems Typo Squat Report](https://user-images.githubusercontent.com/605406/80552749-93248d00-8995-11ea-8d3c-e71f4ed7eac0.png)

![Pie Chart of -_ Confusion](https://user-images.githubusercontent.com/605406/80552751-961f7d80-8995-11ea-87f3-23fb1ab20da4.png)

eliotsykes committed:
Send email to gem owners on yank

Email gem owners each time a version is yanked. The intention is to help
owners detect compromised accounts and attacker-supplanted gems sooner.

sonalkr132 committed:
Limit select of distinct rubygems in just_updated

Previously, this was scanning the complete versions table which was
unnecessary when we only wanted to show 50 versions order by created_at.
Adding limit and order by with agg on created_at in subquery would
ensure we selecting over relevant rows only.

query plan:
https://gist.github.com/sonalkr132/d42a2eac814d9d0e470aa97a94dccede

//before
$ time curl http://localhost:3000/api/v1/activity/just_updated.json -o
/dev/null
  % Total    % Received % Xferd  Average Speed   Time    Time     Time
Current
                                 Dload  Upload   Total   Spent    Left
Speed
100 56324  100 56324    0     0  26381      0  0:00:02  0:00:02 --:--:--
26381

real    0m2.162s
user    0m0.026s
sys     0m0.004s

//after
$ time curl http://localhost:3000/api/v1/activity/just_updated.json -o
/dev/null
  % Total    % Received % Xferd  Average Speed   Time    Time     Time
Current
                                 Dload  Upload   Total   Spent    Left
Speed
100 56324  100 56324    0     0  68188      0 --:--:-- --:--:-- --:--:--
68106

real    0m0.838s
user    0m0.007s
sys     0m0.006s

sonalkr132 committed:
Include gem_downloads to remove n+1 for versions.downloads_count

rietta committed:
Update GemTypo to use the -/_ variation detection

This update changes the method used by GemTypo from the Levenshtein distance
comparison with popular Gems with 10 million or more downloads to a simpler
check that compares the proposed Gem with all of its - and _ characters removed
to all existing Gems with their - and _ characters removed.

This comparison is performed completely within PostgreSQL and an index is added
to ensure that a table scan is not neccesary. This means that this typo protection
can be enforced againstn any other already published Gem rather than just popular
ones with 10 million or more downloads.

My April 2020 analysis of the typo squatting attack in February of 2020 found that
none of the target Gems were protected at the time and the vast majority of the
malicious Gem names would have been prevented by this check if it had been in place
at the time. Please see https://speakerdeck.com/rietta/securing-the-open-source-software-supply-chain
slides 22 and 23 for the publication of my findings at this time.

Fix same name error

This update fixes the error where a gem that is being upgraded was
considered a typo squat. Additionally, I simplified the protected_typo?
method a bit and ensured that only one query was neccesary rather than
two without overfetching data.

Refactor GemTypo tests

Updates to index based on internal code review and stats gathering

Fix broken test

Allow update gem pre-existing Gem that is a typo variant of another

For existing data, where there may be pre-existing typo variants such that
awesome_gem and awesome-gem both already exist. Then skip typo checking if
the current gem name matches either. This does not reduce security in that
other validations already prevent a user from publishing a Gem that has the
exact same name as a Gem already published by another user.

Consider yanked Gems when determining

This update considers the yanked_at status for potential exact name
matches such that a Gem that has at least one not-yanked version would
be considered an exact match. However, a Gem that has all versions yanked
and is also a typo squat of another Gem would be determined to be a
typo squat.

Update GemTypo to use existing case insensitive name index

This update removes the unnecsary downcasing of the target Gem name
and uses the existing UPPER(name) index on the rubygems table to check
for exact name matches.

rietta commented:
This is my internal branch for review.

rietta commented:
Does anything in this pull request appear to be at odds with 

Be sure to review the contributing guidelines and code of conduct. 
https://github.com/rubygems/rubygems.org/blob/master/CONTRIBUTING.md
https://github.com/rubygems/rubygems.org/blob/master/CODE_OF_CONDUCT.md

cchoi12 commented on test/functional/api/v1/rubygems_controller_test.rb:
```diff
@@ -309,19 +309,19 @@ def self.should_respond_to(format)
-    # end
+    context "On POST to create with a protected gem name" do
+      setup do
+        # above_downloads_thres = GemTypo::DOWNLOADS_THRESHOLD + 1
```
Couldn't you remove this?
```suggestion
```

cchoi12 commented:
Also, I reviewed the Contributing MD and Code of Conduct. This PR LGTM :shipit:

rietta commented:
Need to update to allow the name to match if the gem name actually matches, which means it is an update.

rietta committed:
Adjust styling based on RuboCop

sonalkr132 committed:
Increase limit on version metadata value and add validation for key size

This matches allowed size for key and value in gem cli.
Updated tests to check for error message. Previously,
`should "be invalid with maxiumum length"` was raising invalid record
for home url being invalid.

sonalkr132 committed:
Add max length validation to string columns populated by user input

This would have raised server error `ActiveRecord::ValueTooLong
(PG::StringDataRightTruncation: ERROR:  value too long for type
character varying(255))` for values longer than 255 characters.

rails use to set 255 character limit on string columns but not anymore.
all our string columns (in prod) have this limit except `required_rubygems_version`.

sonalkr132 committed:
Add limit to required_rubygems_version

simi committed:
Flag Rubygem as indexed when any version is indexed.

simi committed:
Use Rubygem indexed column to lookup for gems with versions.

This makes lookup much faster:

Version.indexed.distinct.count(:rubygem_id)
   (519.5ms)  SELECT COUNT(DISTINCT versions.rubygem_id) FROM versions WHERE versions.indexed =   [[indexed, true]]
=> 158418

Rubygem.where(indexed: true).count
   (58.4ms)  SELECT COUNT(*) FROM rubygems WHERE rubygems.indexed =   [[indexed, true]]
=> 158418

rietta committed:
Remove ', limit: 255' from schema

Running rake db:migrate had added these limits to the schema automatically. This reverts them.

rietta committed:
Create dashunderscore_typos_idx in background

This update at the suggestion of @simi, asks PostgreSQL to create the new dashunderscore_typos_idx
index in the background without blocking. This will allow the migration to run quickly and cause
no downtime.

simi committed:
Rely on version hook update.

sonalkr132 committed:
Add regex validation to user email

Fixes in delayed job to deliver mails:
```
501 Recipient syntax error

/usr/local/lib/ruby/2.5.0/net/smtp.rb:969:in `check_response'
/usr/local/lib/ruby/2.5.0/net/smtp.rb:937:in `getok'
/usr/local/lib/ruby/2.5.0/net/smtp.rb:865:in `rcptto'
/usr/local/lib/ruby/2.5.0/net/smtp.rb:846:in `block in rcptto_list'
```

sonalkr132 committed:
Remove version number from reverse dependencies page subtitle

The page doesn't show reserve dependencies for the latest version
for the gem, it is confusion to show latest version in the subtitle.

sonalkr132 committed:
Update gem yank to notify notifiable_owners and update mail template

DavisC0801 committed:
Match schema with current master schema

DavisC0801 committed:
Re-add manual gem typo exception whitelist

DavisC0801 committed:
Remove unneeded Gem::Text import statement

DavisC0801 committed:
Add more descriptive error message when names are similar

DavisC0801 committed:
Remove additional comments

rietta committed:
Better explanation for any exact match test

DavisC0801 committed:
Add dash/underscore index to rubygems schema

rietta committed:
Refactor database queries for exceptions and published Gem names

1. Refactor published_exact_name_matches to use existing Scope
2. Updated GemTypoException check to use database instead of array search

rietta committed:
Merge branch 'rietta/typo-squatter-protection' of github.com:rietta/rubygems.org into rietta/typo-squatter-protection

rietta committed:
Adjust matches to Ruby gems that are currently indexed

DavisC0801 committed:
Add test for allowing identically and similarly named gems to be added after a gem is deleted

rietta committed:
Merge branch 'master' into rietta/typo-squatter-protection

rietta commented:
Noticed that CI was failing. I've just merge in the changes from the upstream master and pushed.

rietta committed:
Updates as per Rubocop

DavisC0801 committed:
Add test gem version to rubygem controller test

DavisC0801 committed:
Merge branch 'rietta/typo-squatter-protection' of github.com:rietta-inc/rubygems.org into rietta/typo-squatter-protection

DavisC0801 committed:
Add check for null yanked_at time

DavisC0801 committed:
Refactor test to use versions factory and fix rubocop errors

DavisC0801 committed:
Add existing typosquat setup for any exact match test

indirect committed:
Rescue the versioned "no such file" error from S3

It turns out that when the bucket is versioned, and the latest version
of that key is a deletion marker, the error raised is not `NoSuchKey`,
but instead `NotFound`. I needed to apply this patch live in production
to successfully run `Deletion#restore!` for this ticket:
https://github.com/rubygems/rubygems.org/issues/2400

indirect committed:
Merge pull request #2402 from rubygems/fix-restore

Rescue the versioned "no such file" error from S3

sonalkr132 committed:
Add subtitle_block to layout to move subtitle to new line

sonalkr132 committed:
Add record to track rubygem total downloads

80 rubygems seem to missing following record used to track total
downloads of a rubygem:
```
GemDownload.where(rubygem_id: rubygem.id, version_id: 0)
```
All of these gems have created at date before 2012, while we started
adding above record in 2016. We may have missed adding rubygem record
in backfill because they didn't to have any versions at the time.

We need to create rubygems record for unindexed rubygems as well,
since someone can start using these names in future. for example:
rexml was reserved name but now it is being used.

sonalkr132 committed:
Merge pull request #2346 from sonalkr132/rails-validations

Add validation for max length on string columns with user input

sonalkr132 committed:
Add validation to metadata key being empty

Elasticsearch indexing is breaking as it does allow empty key names
(`metatadata.<key_name>`). ex gem: randomsnumbers
Fixes:
```
[400] {"error":{"root_cause":[{"type":"mapper_parsing_exception","reason":"failed to parse"}],"type":"mapper_parsing_exception","reason":"failed to
parse","caused_by": {"type":"illegal_argument_exception","reason":"name cannot be empty string"}},"status":400}
/usr/local/bundle/gems/elasticsearch-transport-5.0.5/lib/elasticsearch/transport/transport/base.rb:202:in `__raise_transport_error'
/usr/local/bundle/gems/elasticsearch-transport-5.0.5/lib/elasticsearch/transport/transport/base.rb:319:in `perform_request'
/usr/local/bundle/gems/elasticsearch-transport-5.0.5/lib/elasticsearch/transport/transport/http/faraday.rb:20:in `perform_request'
/usr/local/bundle/gems/elasticsearch-transport-5.0.5/lib/elasticsearch/transport/client.rb:131:in `perform_request'
/usr/local/bundle/gems/elasticsearch-api-5.0.5/lib/elasticsearch/api/actions/index.rb:100:in `index'
/usr/local/bundle/gems/elasticsearch-model-5.0.2/lib/elasticsearch/model/indexing.rb:346:in `index_document'
/app/app/models/concerns/rubygem_searchable.rb:9:in `index_document'
```

sonalkr132 committed:
Merge pull request #2389 from sonalkr132/email-regex-validation

Add regex validation to user email

sonalkr132 committed:
Merge pull request #2393 from sonalkr132/gem-yanked-mail

Send email to owners on gem yank

sonalkr132 committed:
Merge pull request #2394 from sonalkr132/rev-dep-version

Remove version number from reverse dependencies page subtitle

sonalkr132 committed:
Increase gem push per hour limit

150 push/hr limit was added when we found issue with exp-backoff calc.
Since then we have fixed issue with exp-backoff calc and updated push
exp-backoff to only rate limit failed requests.
since the back off rate limit is only on failed push requests, this per
hour limit on *all* push requests is still useful. Our jobs queue
gets backup when we see increased push requests.

sonalkr132 committed:
Add test for rubygem indexed on push and yank

sonalkr132 committed:
Merge pull request #2351 from simi/cache-indexed-status-on-rubygem

Cache any indexed version state on Rubygem model.

sonalkr132 committed:
Add script and mail for resetting api keys

Resolves: https://hackerone.com/reports/678037

dependabot[bot] committed:
Bump rack from 2.2.2 to 2.2.3

Bumps [rack](https://github.com/rack/rack) from 2.2.2 to 2.2.3.
- [Release notes](https://github.com/rack/rack/releases)
- [Changelog](https://github.com/rack/rack/blob/master/CHANGELOG.md)
- [Commits](https://github.com/rack/rack/compare/v2.2.2...2.2.3)

Signed-off-by: dependabot[bot] <support@github.com>

Commit:
Merge pull request #2409 from rubygems/dependabot/bundler/rack-2.2.3

Bump rack from 2.2.2 to 2.2.3

DavisC0801 committed:
Update rubygem model to make typosquat error more descriptive

DavisC0801 committed:
Remove unused id select, not used in existing code

DavisC0801 committed:
Added addional tests for missing -/_ and clarified deleted gem tests

sonalkr132 committed:
Merge pull request #2407 from sonalkr132/increase-push-limit

Increase gem push per hour limit

sonalkr132 committed:
Merge pull request #2406 from sonalkr132/metadata-validation

Add validation to metadata key being empty

sonalkr132 committed:
Merge pull request #2404 from sonalkr132/rubygems-gem-downloads

Add record to track rubygem total downloads

sonalkr132 committed:
Merge pull request #2333 from sonalkr132/just-updated-limit

Add limit to just_updated subquery

DavisC0801 committed:
Update schema version to match current master

DavisC0801 committed:
Merge branch 'master' into rietta/typo-squatter-protection

DavisC0801 committed:
Remove whitespace as per rubocop suggestion

DavisC0801 committed:
Update error message for similar gem names

DavisC0801 reviewed (approved):
These changes were already merged into the main branch for rubygems to the best of my knowledge.
